lib/sync/fetcher.go

```diff
@@ -0,0 +1,256 @@
+		nodeAddrs = si.NodeAddrs
+	)
+	f.logger.Debug("Fetch start", "height", height)
+
+	n := f.pickRandomNode(nodeAddrs)
+	if n == nil {
+		return errors.New("Fetch: node not found")
+	}
+	f.logger.Info(fmt.Sprintf("fetching items from node: %v", n), "fetching_node", n, "height", height)
+
+	apiURL := apiClientURL(n, height)
+	f.logger.Debug("apiClient", "url", apiURL.String())
+
+	req, err := http.NewRequest("GET", apiURL.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancelF := context.WithTimeout(ctx, f.fetchTimeout)
+	defer cancelF()
+
+	resp, err := f.apiClient.Do(req)
+	defer func() {
+		if resp != nil && resp.Body != nil {
+			resp.Body.Close()
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		//TODO:
+		err := errors.New("fetch: block not found")
+		return err
+	}
+
+	items, err := f.unmarshalResp(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	//TODO(anarcher): loggger.Debug insteads .Info
+	f.logger.Info("fetch get items", "items", len(items), "height", height)
+
+	blocks, ok := items[runner.NodeItemBlock]
+	if !ok || len(blocks) <= 0 {
+		err := errors.New("fetch: block not found in resp")
+		return err
+	}
+
+	//TODO(anarcher): check items
+	bts, ok := items[runner.NodeItemBlockTransaction]
+
+	blk := blocks[0].(block.Block)
+	si.Block = &blk
+
+	txmap := make(map[string]*transaction.Transaction) // For ordering txs by block.Transactions
+
+	for _, bt := range bts {
+		bt, ok := bt.(block.BlockTransaction)
+		if !ok {
+			//TODO(anarcher): define sential error
+			return errors.New("invalid block transaction")
+		}
+
+		var tx transaction.Transaction
+		if err := json.Unmarshal(bt.Message, &tx); err != nil {
+			return err
+		}
+		txmap[bt.Hash] = &tx
+	}
+
+	for _, hash := range blk.Transactions {
+		tx, ok := txmap[hash]
+		if !ok {
+			//TODO(anarcher): Error type for controlling timeout
+			err := fmt.Errorf("Tx: %s not found in block height %d", hash, height)
+			return err
+		}
+		si.Txs = append(si.Txs, tx)
+
+	}
+
+	return nil
+}
+
+// pickRandomNode choose one node by random. It is very protype for choosing fetching which node
+func (f *BlockFetcher) pickRandomNode(nodeAddrs []string) node.Node {
+	ac := f.connectionManager.AllConnected()
+	if len(ac) <= 0 {
+		return nil
+	}
+
+	var nodeMap = make(map[string]struct{})
+	for _, addr := range nodeAddrs {
+		nodeMap[addr] = struct{}{}
+	}
+
+	var addressList []string
+	for _, a := range ac {
+		if f.localNode.Address() == a {
+			continue
+		}
+		if len(nodeAddrs) > 0 {
+			if _, ok := nodeMap[a]; ok {
+				addressList = append(addressList, a)
+			}
+		} else {
+			addressList = append(addressList, a)
+		}
+	}
+
+	if len(addressList) <= 0 {
+		return nil
+	}
+
+	idx := rand.Intn(len(addressList))
+	node := f.connectionManager.GetNode(addressList[idx])
+	return node
+}
+
+func (f *BlockFetcher) existsBlockHeight(height uint64) bool {
+	exists, err := block.ExistsBlockByHeight(f.storage, height)
```

Comment:
Could we just compare the height to `LatestBlock.Height` ?

Reply:
LastedtHeight  makes sense but Existsblock doesn’t make sense?

Reply:
I'm not sure I understand the question here. Why do you think it doesn't make sense ?

Reply:
ping @Geod24

Reply:
Once again, if you feel strongly about it, this approach is good enough for me.
I just felt we could optimize it by comparing to the latest block (which is more likely to be hot in the cache) than doing a fetch of a specific block.

Reply:
@Geod24 I know your idea. and This is also my idea. I think it is advanced feature for it which I will do it after this PR is merged.  Can you agree this approach? @Geod24 
> but with current design, I think that current code is simple.

Related: #545 (`SyncBlockNotifier`)

Reply:
ping @Geod24

Reply:
@anarcher:
> Once again, if you feel strongly about it, this approach is good enough for me.

So yeah, let's leave it as is at the moment and revisit later!

Reply:
@Geod24 thanks! :->  Did you decide X or O or Need more times with this ugly PR ;ㅁ;

Reply:
It's 2200 lines, I need a bit more time.